<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [380]:
import pandas as pd
import psycopg2
import plotly
import plotly.express as px
import numpy as np
import plotly.graph_objects as go

In [382]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [25]:
query_3_1 = f'''SELECT 
    COUNT(id)
FROM   
    public.vacancies
'''

In [26]:
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_4332\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [28]:
query_3_2 = f'''SELECT 
    COUNT(id)
FROM   
    public.employers
'''

In [29]:
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_4332\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [30]:
query_3_3 = f'''SELECT 
    COUNT(id)
FROM   
    public.areas
'''

In [31]:
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_4332\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [32]:
query_3_4 = f'''SELECT 
    COUNT(id)
FROM   
    public.industries
'''

In [33]:
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_4332\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

Из предварительного анализа данных видим, что таблицы взаимосвязаны между собой по столбцу id, это позволит в дальнейшем выявлять взаимосвязи и проводить анализ данных. В базе 49197 вакансий, 23501 - работодателей, в соотношении - на 1 работодателя приходится по 2 вакансии. В выборке представлены 1362 регинов и 294 сфреы деятельности.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [290]:
query_4_1 = f'''SELECT 
    COUNT(v.id) as cnt,
    a.name as area 
FROM   
    public.vacancies as v
JOIN public.areas as a ON v.area_id = a.id
GROUP BY area
ORDER BY COUNT(v.id) DESC
LIMIT 5    
'''

In [291]:
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_1524\347208320.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt             area
0  5333           Москва
1  2851  Санкт-Петербург
2  2112            Минск
3  2006      Новосибирск
4  1892           Алматы

In [ ]:
fig_1 = px.pie(data_frame=df,                                            
             names='area',                                                   
             values='cnt',                                                      
             labels='area',                                                  
             title='ТОП-5 регионов по количеству вакансий',  
             width=600,                                                           
             height=500)                                                                         
fig_1.show()

In [ ]:
![](images/image_1.png)

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [56]:
query_4_2 = f'''SELECT 
    COUNT(id) as cnt
FROM   
    public.vacancies
    WHERE (salary_from IS NOT NULL) or (salary_to IS NOT NULL)  
'''

In [57]:
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_4332\2168503709.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


cnt
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [58]:
query_4_3 = f'''SELECT 
    round(AVG(salary_from)) as lower_limit,
    round(AVG(salary_to)) as upper_limit
FROM   
    public.vacancies  
'''

In [60]:
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_4332\2222372203.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


lower_limit  upper_limit
0      71065.0     110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [68]:
query_4_4 = f'''SELECT
    schedule,
    employment,
    COUNT(id) as schedule_employment   
FROM   
    public.vacancies
GROUP BY schedule, employment
ORDER BY COUNT(id) DESC    
'''

In [69]:
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_4332\2981602351.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  schedule_employment
0        Полный день     Полная занятость                35367
1   Удаленная работа     Полная занятость                 7802
2      Гибкий график     Полная занятость                 1593
3   Удаленная работа  Частичная занятость                 1312
4     Сменный график     Полная занятость                  940
5        Полный день           Стажировка                  569
6     Вахтовый метод     Полная занятость                  367
7        Полный день  Частичная занятость                  347
8      Гибкий график  Частичная занятость                  312
9        Полный день     Проектная работа                  141
10  Удаленная работа     Проектная работа                  133
11     Гибкий график           Стажировка                  116
12    Сменный график  Частичная занятость                  101
13  Удаленная работа           Стажировка                   64
14     Гибкий график     Проектная работа                   18
15    Сменный график           Стажировка                   12
16    Вахтовый метод     Проектная работа                    2
17    Сменный график     Проектная работа                    1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [279]:
query_4_5 = f'''SELECT
    experience,
    COUNT(id)
FROM   
    public.vacancies
GROUP BY experience
ORDER BY COUNT(id)        
'''

In [280]:
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_1524\2601166264.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

In [ ]:
fig_2 = px.bar(
    data_frame=df,
    x='experience',
    y='count',
    title='Доля вакансий в зависимости от опыта работы'
)
fig_2.show()

In [ ]:
![](images/image_2.png)

***

В ходе детального анализа данных было выявлено, что спрос на персонал наиболее высокий в городах миллионниках, топ-5 по убыванию составили города: Москва, Санкт-Петербург, Минск, Новосибирск, Алматы. Средний размер заработной платы находится на уровне от 71 до 110 тысяч. Наибольшее количество запросов представлено для вакансий с полной занятостью и полным рабочим днем, а также опытом  работы от 1 года до 3 лет, что логично, чем больше опыт, тем ценне становится сотрудник (на молодых специалистов запросов и предложений в работе гораздо больше). Около половины вакансий из выборки представлены с зарплатными ожиданиями.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [295]:
query_5_1 = f'''SELECT
    e.name,
    COUNT(v.id)
FROM   
    public.vacancies v
JOIN public.employers e ON v.employer_id = e.id
GROUP BY e.name
ORDER BY COUNT(v.id) DESC
LIMIT 5
'''

In [296]:
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_1524\1792814825.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

In [ ]:
fig_3 = px.pie(data_frame=df,                                          
             names='name',                                           
             values='count',                                         
             labels='name',                                 
             title='Доля вакансий в ТОП-5 работодателей',            
             width=600,                                              
             height=500)                                             
fig_3.show()

In [ ]:
![](images/image_3.png)

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [276]:
query_5_2 = f'''SELECT 
    a.name as region_name,
    COUNT(distinct e.id) as cnt_employers,
    COUNT(distinct v.id) as cnt_vacancies
FROM   
    public.areas a
JOIN public.employers e ON e.area = a.id
LEFT JOIN public.vacancies v ON v.area_id = a.id
WHERE v.id is NULL
GROUP BY a.name
ORDER BY COUNT(distinct e.id) DESC
LIMIT 5
'''

In [277]:
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_1524\3341481329.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



region_name  cnt_employers  cnt_vacancies
0              Россия            410              0
1           Казахстан            207              0
2  Московская область             75              0
3  Краснодарский край             19              0
4            Беларусь             18              0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [20]:
query_5_3 = f'''SELECT 
    e.name as employer,
    COUNT(v.id) as cnt_vacancies,
    COUNT(distinct a.id) as cnt_region
FROM   
    public.employers e
JOIN public.vacancies v ON e.id = v.employer_id
LEFT JOIN public.areas a ON v.area_id = a.id
GROUP BY 1 
ORDER BY 3 DESC
LIMIT 5
'''

In [21]:
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_1524\2398953733.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


Работодатель  Количество вакансий  Количество регионов
0                  Яндекс                 1933                  181
1              Ростелеком                  491                  152
2              Спецремонт                  151                  116
3  Поляков Денис Иванович                  127                   88
4               ООО ЕФИН                    71                   71

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [43]:
query_5_4 = f'''SELECT 
    COUNT(distinct e.id) as cnt_employers
FROM   
    public.employers e
LEFT JOIN public.employers_industries ei ON e.id = ei.employer_id
WHERE ei.industry_id IS NULL
'''

In [44]:
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_1524\3114411492.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


Количество работодателей
0                      8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [55]:
query_5_5 = f'''SELECT 
    e.name as company_name,
    COUNT(ei.industry_id) as cnt_industry_area
FROM   
    public.employers e
LEFT JOIN public.employers_industries ei ON e.id = ei.employer_id
GROUP BY e.name
HAVING COUNT(ei.industry_id) = 4
ORDER BY e.name
OFFSET 2
LIMIT 1
'''

In [56]:
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_1524\1650419958.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


Название компании  Количество сфер деятельности
0              2ГИС                             4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [67]:
query_5_6 = f'''SELECT 
    COUNT(e.id) as cnt_employers
FROM   
    public.employers e
LEFT JOIN public.employers_industries ei ON e.id = ei.employer_id
JOIN public.industries i ON i.id = ei.industry_id
WHERE i.name = 'Разработка программного обеспечения'
'''

In [68]:
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_1524\1169039133.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


cnt_employers
0           3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [82]:
from urllib.parse import quote_plus
url = quote_plus('https://ru.wikipedia.org/wiki/Города-миллионеры_России', safe=':/')
tables = pd.read_html(url)
mln_cities = tables[1]['Город'].values.tolist()
print(mln_cities)

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград']


In [87]:
query_5_7 = f'''SELECT 
    a.name as region_mln,
    count(v.id) as cnt_vacancies
FROM
    public.vacancies as v
JOIN public.employers as e on v.employer_id = e.id
JOIN public.areas as a on v.area_id = a.id
GROUP BY v.area_id, a.name, e.id, e.name
HAVING a.name in {tuple(mln_cities)} and e.name = 'Яндекс'

UNION ALL

SELECT
    'Total',
    SUM(x.cnt)
    FROM
        (
        select
            count(*) as cnt
        from
            public.vacancies as v
            join public.employers as e on v.employer_id = e.id
            join public.areas as a on v.area_id = a.id
        group by
            v.area_id, a.name, e.id, e.name
        having
            a.name in {tuple(mln_cities)} and e.name = 'Яндекс'
        ) as x
ORDER BY 2 
'''

In [88]:
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_1524\3129374555.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


region_mln  cnt_vacancies
0              Омск           21.0
1         Челябинск           22.0
2        Красноярск           23.0
3         Волгоград           24.0
4    Ростов-на-Дону           25.0
5            Казань           25.0
6             Пермь           25.0
7               Уфа           26.0
8            Самара           26.0
9         Краснодар           30.0
10          Воронеж           32.0
11      Новосибирск           35.0
12  Нижний Новгород           36.0
13     Екатеринбург           39.0
14  Санкт-Петербург           42.0
15           Москва           54.0
16            Total          485.0

***

Из анализа работодателей можем сделать вывод, наибольшее количество ваканисий представлено в регионах-миллионниках, повышенный спрос на ресурсы наблюдается в сфере IT, телекоммуникаций, финансовой и нефтяной области. Также присутствуют вакансии с без указания сферы деятельности.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [113]:
query_6_1 = f'''SELECT 
    COUNT(name) as cnt_vacancies
FROM   
    public.vacancies 
WHERE name ILIKE '%data%'
    OR name ILIKE '%данн%'
'''

In [114]:
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_1524\485020324.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


cnt_vacancies
0           1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [166]:
query_6_2 = f'''SELECT 
    COUNT(*) as cnt_vacancies
FROM   
    public.vacancies v 
WHERE (v.name ILIKE '%data scientist%'
    OR v.name ILIKE '%data science%'
    OR v.name ILIKE '%исследователь данных%'
    OR (v.name ILIKE '%ML%' AND v.name NOT ILIKE '%HTML%')
    OR v.name ILIKE '%machine learning%'
    OR v.name ILIKE '%машинн%обучен%')
    AND 
    (v.name ILIKE '%junior%'
    OR v.experience ILIKE '%Нет опыта%' 
    OR v.employment ILIKE '%Стажировка%')
'''

In [167]:
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_1524\3641650667.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


cnt_vacancies
0             51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [179]:
query_6_3 = f'''SELECT 
    COUNT(*) as cnt_vacancies
FROM   
    public.vacancies v 
WHERE (v.name ILIKE '%data scientist%'
    OR v.name ILIKE '%data science%'
    OR v.name ILIKE '%исследователь данных%'
    OR (v.name ILIKE '%ML%' AND v.name NOT ILIKE '%HTML%')
    OR v.name ILIKE '%machine learning%'
    OR v.name ILIKE '%машинн%обучен%')
    AND 
    (key_skills ILIKE '%SQL%'
    OR key_skills ILIKE '%postgres%')
'''

In [180]:
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_1524\1159587760.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


cnt_vacancies
0            229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [254]:
query_6_4 = f'''SELECT 
    COUNT(*) as cnt_vacancies
FROM   
    public.vacancies v 
WHERE (v.name ILIKE '%data scientist%'
    OR v.name ILIKE '%data science%'
    OR v.name ILIKE '%исследователь данных%'
    OR (v.name ILIKE '%ML%' AND v.name NOT ILIKE '%HTML%')
    OR v.name ILIKE '%machine learning%'
    OR v.name ILIKE '%машинн%обучен%')
    AND 
    key_skills ILIKE '%Python%'
'''

In [255]:
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_1524\972386973.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


cnt_vacancies
0            357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [256]:
query_6_5 = f'''SELECT 
    round(avg(length(key_skills)-length(replace(key_skills, '\t', ''))+1), 2) as result
FROM   
    public.vacancies v 
WHERE (v.name ILIKE '%data scientist%'
    OR v.name ILIKE '%data science%'
    OR v.name ILIKE '%исследователь данных%'
    OR (v.name LIKE '%ML%' AND v.name NOT ILIKE '%HTML%')
    OR v.name ILIKE '%machine learning%'
    OR v.name ILIKE '%машинн%обучен%')
'''

In [257]:
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_1524\390248827.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


result
0    6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [266]:
query_6_6 = f'''SELECT
    experience,
    round(avg(coalesce((v.salary_from + salary_to)/2, salary_from, salary_to)), 0) as result
FROM   
    public.vacancies v 
WHERE (v.name ILIKE '%data scientist%'
    OR v.name ILIKE '%data science%'
    OR v.name ILIKE '%исследователь данных%'
    OR (v.name LIKE '%ML%' AND v.name NOT ILIKE '%HTML%')
    OR v.name ILIKE '%machine learning%'
    OR v.name ILIKE '%машинн%обучен%')
    AND
    (v.salary_from IS NOT NULL 
    OR v.salary_to IS NOT NULL)
    AND 
    experience ILIKE 'От 3 до 6 лет'
GROUP BY 1
'''

In [267]:
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_1524\3561045109.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience    result
0  От 3 до 6 лет  243115.0

***

Говоря о предметном анализе, можно сделать вывод, что вакансий представленных в области ДС совсем не много (1771), а для начинающих специалистов в ДС их всего 51. Также к специалистам ДС предъявляются требования, кандидат должен владеть как минимум 6 навыками, часто встречающиеся требования: язык программирования Python и владение SQL или postgres.
Средняя заработная плата для DS с оптытом работы от 3 до 6 лет составляет 243 тр. Вакансий для ведущих специалистов с опытом работы старше 6 лет нет совсем, что может говорить о том, что работодатели заинтересованы в ДС с опытом работы, либо миддл и синьор-специалисты справляются с поставленными задачами.  

# Общий вывод по проекту

Беря во внимание проведенный ранее анализ данных можно сделать выводы:
1. достаточно небольшое количество работодателей прибегают к поиску специалистов ДС;
2. прослеживается тенденция к бурному развитию в сфере IT (об этом нам говорит большой спрос на вакансии в данной сфере);
3. ожидаемо выше спрос на вакансии любой сферы в Москве и прочих городах миллионниках (спрос на специалистов сферы ДС в т.ч.);
4. вакансии для специалистов сферы ДС предлагают крупные стабильные Российские компании, также можно расчитывать и на предложения в Западных компаниях.
5. знания и навыки сотрудников ДС широко используются в современной действительности.Такие секторы как финансовый, рекрутинговый, телекоммуникационный находятся в поиске специалистов ДС.

#здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

7.1 Узнаем самую высокую границу заработной платы (salary_from). Выведем ТОП-5 самых "дорогих" вакансий опубликованных в выборке с указанием уровня зарплат. 

In [351]:
query_7_1 = f'''SELECT 
    name,
    salary_from
FROM   
    public.vacancies
WHERE salary_from IS NOT NULL
ORDER BY 2 DESC
LIMIT 5
'''

In [352]:
df = pd.read_sql_query(query_7_1, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_1524\2911634181.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  salary_from
0                               Технический директор      1000000
1             Директор по информационным технологиям       650000
2              Frontend Developer in Unicorn Company       574000
3             FullStack Developer in Unicorn Company       574000
4  Java-разработчик (со 2-ой формой допуска к гос...       500000

7.2 Посчитаем количество вакансий с заработными платами до 150, от 150 до 250, от 250 до 350, от 350 до 450, выше 450 (верхняя граница условий не включается в расчет: (0; 150), `[150, 250), [250, 350), и т.д.)
Для расчета берем вакансии в которых указана хотя бы одна граница уровня заработной платы.

In [355]:
query_7_2 = f'''SELECT 
    COUNT(id) cnt_vacancies, 'до 150 тыс' salary
FROM   
    public.vacancies
WHERE salary_from IS NOT NULL and salary_from < 150000

UNION ALL

SELECT
    COUNT(id) cnt_vacancies, 'от 150 до 250 тыс' salary
FROM   
    public.vacancies
WHERE salary_from IS NOT NULL and salary_from >= 150000 AND salary_from < 250000

UNION ALL

SELECT
    COUNT(id) cnt_vacancies, 'от 250 до 350 тыс' salary
FROM   
    public.vacancies
WHERE salary_from IS NOT NULL and salary_from >= 250000 AND salary_from < 350000

UNION ALL

SELECT
    COUNT(id) cnt_vacancies, 'от 350 до 450 тыс' salary
FROM   
    public.vacancies
WHERE salary_from IS NOT NULL and salary_from >= 350000 AND salary_from < 450000

UNION ALL

SELECT
    COUNT(id) cnt_vacancies, 'от 450 тыс' salary
FROM   
    public.vacancies
WHERE salary_from IS NOT NULL and salary_from >= 450000

UNION ALL

SELECT
    COUNT(id) cnt_vacancies, 'ЗП не указана' salary
FROM   
    public.vacancies
WHERE salary_from IS NULL

ORDER BY 1 DESC
'''

In [356]:
df = pd.read_sql_query(query_7_2, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_1524\2993136978.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt_vacancies             salary
0          27712      ЗП не указана
1          19142         до 150 тыс
2           1906  от 150 до 250 тыс
3            350  от 250 до 350 тыс
4             66  от 350 до 450 тыс
5             21         от 450 тыс

Как видно из данных, чуть меньше половины из всех представленных вакансий с предлагаемой заработной платой до 150 тысяч. С ростом уровня зп снижается количество предлагаемых вакансий. 

7.3 Узнаем, кто из работодателей чаще всего подыскивает дата-сайентистистов? 

In [374]:
query_7_3 = f'''SELECT
    e.name as employer,
    i.name as industry,
    COUNT(v.id) as cnt_vacancies
FROM   
    public.vacancies as v
JOIN public.employers as e on v.employer_id = e.id
JOIN public.employers_industries as ei on ei.employer_id = e.id
JOIN public.industries as i on i.id = ei.industry_id
WHERE (v.name ILIKE '%data scientist%'
    OR v.name ILIKE '%data science%'
    OR v.name ILIKE '%исследователь данных%'
    OR (v.name LIKE '%ML%' AND v.name NOT ILIKE '%HTML%')
    OR v.name ILIKE '%machine learning%'
    OR v.name ILIKE '%машинн%обучен%')
GROUP BY i.name, e.name
ORDER BY 3 DESC
LIMIT 20
'''

In [375]:
df = pd.read_sql_query(query_7_3, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_1524\2328698441.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



employer                                           industry  \
0                    СБЕР                                               Банк   
1         Bell Integrator  Системная интеграция,  автоматизации технологи...   
2          Банк ВТБ (ПАО)  Управляющая, инвестиционная компания (управлен...   
3          Банк ВТБ (ПАО)                                               Банк   
4                      VK  Маркетинговые, рекламные, BTL, дизайнерские, E...   
5                      VK                Разработка программного обеспечения   
6                      VK  Интернет-компания (поисковики, платежные систе...   
7   Positive Technologies                Разработка программного обеспечения   
8   Positive Technologies  Системная интеграция,  автоматизации технологи...   
9                  Яндекс  Интернет-компания (поисковики, платежные систе...   
10                МегаФон                                Фиксированная связь   
11                МегаФон  Системная интеграция,  автоматизации технологи...   
12                МегаФон                Разработка программного обеспечения   
13                МегаФон                                    Мобильная связь   
14               Andersen                Разработка программного обеспечения   
15               Andersen  Системная интеграция,  автоматизации технологи...   
16     Бэнкс Софт Системс                Разработка программного обеспечения   
17                   Ozon                                   Интернет-магазин   
18                    МТС                                    Мобильная связь   
19                 Контур  Системная интеграция,  автоматизации технологи...   

    cnt_vacancies  
0              37  
1              25  
2              18  
3              18  
4              15  
5              15  
6              15  
7              11  
8              11  
9               9  
10              8  
11              8  
12              8  
13              8  
14              7  
15              7  
16              6  
17              6  
18              6  
19              6

7.4 Узнаем какой наиболее популярный тип рабочего графика и трудоустройства для дата-сайентистов.

In [452]:
query_7_4 = f'''SELECT 
    COUNT(1) cnt_DS_vacancies, t.schedule, t.employment
FROM
(SELECT 
    v.name as cnt_DS_vacancies,
    coalesce(v.schedule,'График не указан') as schedule,
    coalesce(v.employment,'Тип трудоустройства не указан') as employment
FROM   
    public.vacancies v
WHERE (v.name ILIKE '%data scientist%'
    OR v.name ILIKE '%data science%'
    OR v.name ILIKE '%исследователь данных%'
    OR (v.name LIKE '%ML%' AND v.name NOT ILIKE '%HTML%')
    OR v.name ILIKE '%machine learning%'
    OR v.name ILIKE '%машинн%обучен%')


UNION ALL

SELECT 
    v.name as cnt_DS_vacancies,
    'total' as schedule, 
    '' as employment
FROM   
    public.vacancies v
WHERE (v.name ILIKE '%data scientist%'
    OR v.name ILIKE '%data science%'
    OR v.name ILIKE '%исследователь данных%'
    OR (v.name LIKE '%ML%' AND v.name NOT ILIKE '%HTML%')
    OR v.name ILIKE '%machine learning%'
    OR v.name ILIKE '%машинн%обучен%')
) t
GROUP BY t.schedule, t.employment
ORDER BY 1 
'''

In [453]:
df = pd.read_sql_query(query_7_4, connection)
df

C:\Users\mtsul\AppData\Local\Temp\ipykernel_1524\2502122577.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt_ds_vacancies          schedule           employment
0                 1       Полный день  Частичная занятость
1                 2       Полный день           Стажировка
2                 2  Удаленная работа     Проектная работа
3                 2     Гибкий график           Стажировка
4                 3  Удаленная работа  Частичная занятость
5                 4     Гибкий график  Частичная занятость
6                37     Гибкий график     Полная занятость
7               110  Удаленная работа     Полная занятость
8               319       Полный день     Полная занятость
9               480             total

Из проделанного анализа видим, что из 480 вакансий для дата-сайентистов представленных в выборке 319 вакансий с графиком полная занятость на полный рабочий день, второе место по популярности (110 вакансий) - удаленная работа - полная занятость, есть 4 вакансии для стажировки.

In [454]:

connection.close()